File: sdk/python/gladys-sensor-sdk/src/gladys_sensor_sdk/args.py
"""Typed command arguments with lenient parsing.

Each command has a typed args class with named accessors, default values,
and a raw() escape hatch for undocumented fields. Parsing is lenient:
missing fields get defaults, wrong types get defaults, never throws.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True)
class CommandArgs:
    """Base class for all typed command arguments.

    Provides lenient raw() accessor for unknown fields.
    """

    _raw: dict[str, Any] = field(default_factory=dict, repr=False)

    def raw(self, key: str, default: Any = None) -> Any:
        """Get raw argument value (escape hatch for undocumented args).

        Args:
            key: Argument name.
            default: Default value if key not found.

        Returns:
            Argument value or default.
        """
        return self._raw.get(key, default)


def _safe_bool(value: Any, default: bool) -> bool:
    """Coerce value to bool leniently. Returns default on failure."""
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        return value.lower() in ("true", "1", "yes")
    if isinstance(value, (int, float)):
        return bool(value)
    return default


def _safe_int(value: Any, default: int) -> int:
    """Coerce value to int leniently. Returns default on failure."""
    if isinstance(value, int) and not isinstance(value, bool):
        return value
    if isinstance(value, float):
        if not math.isfinite(value):
            return default
        try:
            return int(value)
        except (ValueError, OverflowError):
            return default
    if isinstance(value, str):
        try:
            return int(value)
        except (ValueError, TypeError):
            return default
    return default


def _safe_str(value: Any, default: str) -> str:
    """Coerce value to str leniently. Returns default on failure."""
    if isinstance(value, str):
        return value
    if value is None:
        return default
    return str(value)


@dataclass(frozen=True)
class StartArgs(CommandArgs):
    """Arguments for START command.

    Attributes:
        dry_run: Validate config without starting.
    """

    dry_run: bool = False

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> StartArgs:
        """Parse args from dict (lenient parsing).

        Missing fields get defaults. Wrong types get defaults.
        Never raises exceptions.
        """
        return cls(
            dry_run=_safe_bool(raw.get("dry_run"), False),
            _raw=raw,
        )

    @classmethod
    def test_defaults(cls) -> StartArgs:
        """Factory for tests (normal startup)."""
        return cls()

    @classmethod
    def test_dry_run(cls) -> StartArgs:
        """Factory for tests (dry run mode)."""
        return cls(dry_run=True)


@dataclass(frozen=True)
class StopArgs(CommandArgs):
    """Arguments for STOP command.

    Attributes:
        force: Skip graceful shutdown.
        timeout_ms: Shutdown timeout in milliseconds.
    """

    force: bool = False
    timeout_ms: int = 5000

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> StopArgs:
        """Parse args from dict (lenient parsing).

        Missing fields get defaults. Wrong types get defaults.
        Never raises exceptions.
        """
        return cls(
            force=_safe_bool(raw.get("force"), False),
            timeout_ms=_safe_int(raw.get("timeout_ms"), 5000),
            _raw=raw,
        )

    @classmethod
    def test_defaults(cls) -> StopArgs:
        """Factory for tests (graceful shutdown)."""
        return cls()

    @classmethod
    def test_force(cls) -> StopArgs:
        """Factory for tests (force shutdown)."""
        return cls(force=True, timeout_ms=1000)


@dataclass(frozen=True)
class RecoverArgs(CommandArgs):
    """Arguments for RECOVER command.

    Attributes:
        strategy: Recovery strategy identifier.
    """

    strategy: str = "default"

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> RecoverArgs:
        """Parse args from dict (lenient parsing).

        Missing fields get defaults. Wrong types get defaults.
        Never raises exceptions.
        """
        return cls(
            strategy=_safe_str(raw.get("strategy"), "default"),
            _raw=raw,
        )

    @classmethod
    def test_defaults(cls) -> RecoverArgs:
        """Factory for tests (default strategy)."""
        return cls()


@dataclass(frozen=True)
class HealthCheckArgs(CommandArgs):
    """Arguments for HEALTH_CHECK command.

    Attributes:
        deep: Perform comprehensive health check.
    """

    deep: bool = False

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> HealthCheckArgs:
        """Parse args from dict (lenient parsing).

        Missing fields get defaults. Wrong types get defaults.
        Never raises exceptions.
        """
        return cls(
            deep=_safe_bool(raw.get("deep"), False),
            _raw=raw,
        )

    @classmethod
    def test_defaults(cls) -> HealthCheckArgs:
        """Factory for tests (standard health check)."""
        return cls()

    @classmethod
    def test_deep(cls) -> HealthCheckArgs:
        """Factory for tests (deep health check)."""
        return cls(deep=True)

File: sdk/js/gladys-sensor-sdk/src/args.ts
/**
 * Per-command typed arguments with lenient parsing.
 * Missing fields get defaults, wrong types get defaults, never throws.
 * Use raw() escape hatch for undocumented/sensor-specific fields.
 */

/** Base interface for all command argument types. */
export interface CommandArgs {
  /** Get raw value from args struct. Returns defaultValue if key doesn't exist. */
  raw(key: string, defaultValue?: unknown): unknown;
}

// --- Helpers for lenient parsing ---

function asBool(value: unknown, defaultValue: boolean): boolean {
  if (typeof value === "boolean") return value;
  if (typeof value === "string") {
    if (value.toLowerCase() === "true") return true;
    if (value.toLowerCase() === "false") return false;
  }
  if (typeof value === "number") return value !== 0;
  return defaultValue;
}

function asNumber(value: unknown, defaultValue: number): number {
  if (typeof value === "number" && !isNaN(value)) return value;
  if (typeof value === "string") {
    const parsed = Number(value);
    if (!isNaN(parsed)) return parsed;
  }
  return defaultValue;
}

function asString(value: unknown, defaultValue: string): string {
  if (typeof value === "string") return value;
  if (value !== null && value !== undefined) return String(value);
  return defaultValue;
}

// --- StartArgs ---

/** Arguments for START command. */
export interface StartArgs extends CommandArgs {
  /** If true, validate configuration but don't start. Default: false */
  readonly dryRun: boolean;
}

class StartArgsImpl implements StartArgs {
  readonly dryRun: boolean;
  private readonly fields: Record<string, unknown>;

  constructor(fields: Record<string, unknown>) {
    this.fields = fields;
    this.dryRun = asBool(fields["dryRun"] ?? fields["dry_run"], false);
  }

  raw(key: string, defaultValue?: unknown): unknown {
    return this.fields[key] ?? defaultValue;
  }

  /** Parse from struct with lenient coercion. */
  static fromStruct(struct: Record<string, unknown>): StartArgs {
    return new StartArgsImpl(struct ?? {});
  }

  /** Test factory with all defaults. */
  static testDefaults(): StartArgs {
    return new StartArgsImpl({});
  }

  /** Test factory with dryRun=true. */
  static testDryRun(): StartArgs {
    return new StartArgsImpl({ dryRun: true });
  }
}

/** Parse StartArgs from a plain object. */
export function parseStartArgs(struct: Record<string, unknown> | undefined): StartArgs {
  return StartArgsImpl.fromStruct(struct ?? {});
}

/** Create StartArgs with all defaults (for testing). */
export function startArgsDefaults(): StartArgs {
  return StartArgsImpl.testDefaults();
}

/** Create StartArgs with dryRun=true (for testing). */
export function startArgsDryRun(): StartArgs {
  return StartArgsImpl.testDryRun();
}

// --- StopArgs ---

/** Arguments for STOP command. */
export interface StopArgs extends CommandArgs {
  /** If true, force immediate shutdown without cleanup. Default: false */
  readonly force: boolean;
  /** Shutdown timeout in milliseconds. Default: 5000 */
  readonly timeoutMs: number;
}

class StopArgsImpl implements StopArgs {
  readonly force: boolean;
  readonly timeoutMs: number;
  private readonly fields: Record<string, unknown>;

  constructor(fields: Record<string, unknown>) {
    this.fields = fields;
    this.force = asBool(fields["force"], false);
    this.timeoutMs = asNumber(fields["timeoutMs"] ?? fields["timeout_ms"], 5000);
  }

  raw(key: string, defaultValue?: unknown): unknown {
    return this.fields[key] ?? defaultValue;
  }

  static fromStruct(struct: Record<string, unknown>): StopArgs {
    return new StopArgsImpl(struct ?? {});
  }

  static testDefaults(): StopArgs {
    return new StopArgsImpl({});
  }

  static testForce(): StopArgs {
    return new StopArgsImpl({ force: true });
  }
}

/** Parse StopArgs from a plain object. */
export function parseStopArgs(struct: Record<string, unknown> | undefined): StopArgs {
  return StopArgsImpl.fromStruct(struct ?? {});
}

/** Create StopArgs with all defaults (for testing). */
export function stopArgsDefaults(): StopArgs {
  return StopArgsImpl.testDefaults();
}

/** Create StopArgs with force=true (for testing). */
export function stopArgsForce(): StopArgs {
  return StopArgsImpl.testForce();
}

// --- RecoverArgs ---

/** Arguments for RECOVER command. */
export interface RecoverArgs extends CommandArgs {
  /** Recovery strategy identifier. Default: "default" */
  readonly strategy: string;
}

class RecoverArgsImpl implements RecoverArgs {
  readonly strategy: string;
  private readonly fields: Record<string, unknown>;

  constructor(fields: Record<string, unknown>) {
    this.fields = fields;
    this.strategy = asString(fields["strategy"], "default");
  }

  raw(key: string, defaultValue?: unknown): unknown {
    return this.fields[key] ?? defaultValue;
  }

  static fromStruct(struct: Record<string, unknown>): RecoverArgs {
    return new RecoverArgsImpl(struct ?? {});
  }

  static testDefaults(): RecoverArgs {
    return new RecoverArgsImpl({});
  }
}

/** Parse RecoverArgs from a plain object. */
export function parseRecoverArgs(struct: Record<string, unknown> | undefined): RecoverArgs {
  return RecoverArgsImpl.fromStruct(struct ?? {});
}

/** Create RecoverArgs with all defaults (for testing). */
export function recoverArgsDefaults(): RecoverArgs {
  return RecoverArgsImpl.testDefaults();
}

// --- HealthCheckArgs ---

/** Arguments for HEALTH_CHECK command. */
export interface HealthCheckArgs extends CommandArgs {
  /** If true, perform deep health check. Default: false */
  readonly deep: boolean;
}

class HealthCheckArgsImpl implements HealthCheckArgs {
  readonly deep: boolean;
  private readonly fields: Record<string, unknown>;

  constructor(fields: Record<string, unknown>) {
    this.fields = fields;
    this.deep = asBool(fields["deep"], false);
  }

  raw(key: string, defaultValue?: unknown): unknown {
    return this.fields[key] ?? defaultValue;
  }

  static fromStruct(struct: Record<string, unknown>): HealthCheckArgs {
    return new HealthCheckArgsImpl(struct ?? {});
  }

  static testDefaults(): HealthCheckArgs {
    return new HealthCheckArgsImpl({});
  }
}

/** Parse HealthCheckArgs from a plain object. */
export function parseHealthCheckArgs(struct: Record<string, unknown> | undefined): HealthCheckArgs {
  return HealthCheckArgsImpl.fromStruct(struct ?? {});
}

/** Create HealthCheckArgs with all defaults (for testing). */
export function healthCheckArgsDefaults(): HealthCheckArgs {
  return HealthCheckArgsImpl.testDefaults();
}
